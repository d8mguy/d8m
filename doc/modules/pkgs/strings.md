## Strings

This module provides access to a subset of the functions in the corresponding go package, plus a few others.


#### function split

    val split = \(strg, sep: string) -> list(string) {...}
Be careful: go `Split` doesn't return an empty string if no instances of `sep` in `strg`, it returns a singleton list of strg. Thus, the correct test for no separators is split.count == 1, not 0.

#### function join

    val join = \(strgs: list(string), inter: string) -> string {...}


#### function toLower

    val toLower = \(strg: string) -> string {...}

#### function toUpper

    val toUpper = \(strg: string) -> string {...}

#### function toCap

    val toCap = \(strg: string) -> string {...}

#### function toCamel

    val toCamel = \(strgs: list(string), initial: boolean) -> string {...}
Generate a camel case string from the input strgs; `initial` says whether to capitalize the first word.

#### function genUpper

    val genUpper = \imp(inx: integer) -> string {...}
`Inx` encodes a base-26 number; generate an upper case letter sequence from it.

#### function genLower

    val genLower = \imp(inx: integer) -> string {...}
`Inx` encodes a base-26 number; generate a lower case letter sequence from it.

#### function toOrdinal

    val toOrdinal = \imp(xxx: integer) -> string {...}
This isn't the only way to generate ordinals but it's reasonable.

#### function startsWith

    val startsWith = \(strg, ptn: string) -> boolean {...}
True if `strg` starts with `ptn`.

#### function endsWith

    val endsWith = \(strg, ptn: string) -> boolean {...}
True if `strg` ends with `ptn`.

#### function trimSpace

    val trimSpace = \imp(strg: string) -> string {...}
Trim whitespace from both ends. This is done in a go pkg so we use that.

