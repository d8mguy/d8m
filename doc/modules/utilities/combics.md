## Combics

Combics provides a selection of useful combinatorial algorithms. The idea is to add more definitions over time.

#### enumCombns:function

    val enumCombns : \(n, k, totcount: integer) -> list(list(integer))

`enumCombns` returns a maximum of `totcount` of the combinations of n choose k. For this to make sense, we need `k >= 1` and `k < n`; `enumCombns` panics unless these conditions are met. Each combination is a list(integer) of length k of integers between 0 and n (exclusive); in addition, each combination will have strictly increasing values and the list of combinations is ordered lexicographically.

#### randperm:function

    val randperm : \mod(list(T)) -> list(T) for any T::entity

Randomly permute a list of type `T`, using go package `math/rand` with implicit seed to generate the random numbers. It modifies and returns the list. Note that repeated calls to this function will not generate all permutations of the list without repetition. Note also that since it uses the PRNG state built into `math/rand`, using this may change the behavior of other program elements that also use that PRNG. If this is a problem, you can easily write
a version with its own state; the code is just 8 lines long. The algorithm used here is called the Fisher-Yates shuffle; as explained in the wikipedia, a version was published in 1938 by Fisher and Yates, and popularized in vol 2 of Knuth's The Art of Computer Programming.

#### streamIndices:type

    val streamIndices = extend tuple(cur, limits: list(integer)) with {
        method lithook : \mod(lims: list(integer))
        method eltcount : \() -> integer
        method nDimensions : \() -> integer
        method restart : \mod() -> nothing
        method done = \() -> boolean
        method current = \() -> list(integer)
        method next = \mod() -> list(integer)
    }

A stream type that accepts a list of lengths and generates all the indices up to each of those lengths, in
lexicographic order. Thus, given [1,2,3] it will generate [[0,0,0],[0,0,1],[0,0,2],[0,1,0],[0,1,1],[0,1,2]].
Note that `[streamIndices: ]` is immediately done, it generates nothing. This is the right way to handle the situation where you need to enumerate all the elements of a variable number of lists. It's sensible to think of the k
values that initialize a `streamIndices` as the extents of a k-dimensional cube. From this perspective, the stream generates all vertices of the cube in lexicographic order. A  `streamIndices` will panic if any initializing length
is <= 0 or `next` is called when `done`.

A `stream type` by convention has methods named `current`, `next`, and `done`; you call `next` until `done` becomes true. 