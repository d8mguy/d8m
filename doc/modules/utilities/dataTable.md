## DataTable

Module dataTable lets you read 2d data tables in files or other generators, and do cube projections on the 2d data in memory. The module treats 2d data as `list(list(string))`, in other words data is uninterpreted except for tabular structure. Arbitrary dimensions can be held in a single data table, and related data tables can hold data that add more dimensions. 

A `paramT` defines a dimension for data. So files hold one or more dimensions in their output column and may or may not list the values of the parameters in the files themselves. All this is described in a `fileFormat`, which is held in a `datascheme`. The `fileFormat` describes the format of a single file. Besides the `fileFormat`, a `datascheme` holds a list of "xparams" and a function called dataGetter. The xparams describe the dimensions that are split across files, one element per file. The caller is responsible for assigning how filenames map to these elements, datascheme calls the dataGetter with the filenames supplied by the user and caches the contents. The name "dataGetter" is meant to emphasize that the "filename" can be any sort of descriptor; the source of data could be a URI/URL or anything convenient.

Besides the relevant types and their methods, dataTable exports stdFileGetter which you can use as a dataGetter for local files having a particularly simple format.

    val rowcolT = extend list(list(string)) where {
        method extractParameters = \imp(nparams: integer) -> rowcolT {...}
        method checkParameterEncoding = \imp(nparams: integer) -> string {...}
        method checkCounts = \(filename:string, xR, xC: integer) -> string {...}
    }
Method `extractParameters`: given a number of initial columns that are values for row encoded param, assumed in lexicographic order (column 0 varying slowest), return a `list(list(string))` with outer list indexed by param and inner lists being extracted lists of values. Code assumes nparams > 0.

Method `checkParameterEncoding`: do the same thing as `extractParameters` but check consistency instead of extracting element values. The consistency checking that's done is to verify that the values in the first N columns (N=# of params) repeat in the way dictated by enumerating those parameters lexicographically. Return a string describing the first problem found in the encoding, "" if no problem found. This works a lot harder than extractParameters; call it if you have any doubts about the integrity of your data source.

Method `checkCounts`: this is used when reading a file into the cache. It returns a non-empty string if there is a problem.


    val paramT = tuple(ident, descrn, unit: string, elts: list(string), rowEncoded:boolean, explicit: boolean)
`ParamT` describes the parameters encoded in a file: `ident` suitable as a label and `descrn` as a longer explanation; `unit` optional, `elts` enumerates the values this parameter takes in the file. The two booleans describe whether and how this parameter occurs in the file: if explicit then must be row encoded; if implicit can be encoded in either row or cols (the latter if `rowEncoded` is false). Another constraint on these encodings: all implicit column encoded params must occur after row encoded ones. All these constraints are checked in the `fileFormat` constructor.

    val stdFileGetter = \imp(fnm: string) -> rowcolT {...}
`DataScheme` has a `dataGetter:\(string)->rowcolT` attribute; `stdFileGetter` is a simple example of a suitable function. Given a path to a csv file with lines split on `\n`, no quotation, extra spaces, or extra lines, open, read, and close the file, split and return contents.

    [fielFormat: prms, outs]
where `prms: list(paramT)`, `outs: list(string)`. 

    val fileFormat = extend tuple(params: list(paramT), outputs: list(string)) where {
        method colEncodedCount = \() -> integer {...}
        method firstOutputIndex = \() -> integer {...}
        method nCols = \() -> integer {...}
        method nRows = \() -> integer {...}
    }
A file has `R` rows and `C` cols where `C = params.count + outputs.count`, `R = reduce(params.{elts.count}, $*, 1)`.

    [datascheme: ffmt, dg]
where `ffmt: fileFormat` and `dg:\(string)->rowcolT`

    val datascheme = extend tuple(filefmt: fileFormat, private xtparams: list(paramT), 
        dataGetter: \(string)->rowcolT) where {
        method update = \mod(xtprms:list(paramT), fnms:list(string), fpfx: string) -> nothing {...}
        method extract = \imp(pIndices, outputs: list(integer), colThresh: integer) -> rowcolT {...}
    }
A `datascheme` allows for extending a `fileFormat` across multiple "files" related by "xtparams" (external params); the `dataGetter` parameterizes the process of retrieving a data table. Internal to `datascheme` we manage a caching scheme so that the dataGetter is called at most once for each distinct filename. Note that xtparams is private to encourage you to change it only via `update`. Also, that an earlier design contemplated using a function/streamer to generate the correct sequence of filenames to pass to dataGetter but that in the end it was easier to provide a list of filenames.

Method `update`: changes xtparams and corresponding file related things.

Method `extract`: produces a data table of strings from a spec for extraction from self. The spec comes from 3 args: pIndices is a list of integers where >= 0 means that index in the corresponding param or xtparam and < 0 means the corresponding param or xtparam is enumerated in the output. Negative numbers should all be distinct and contiguous; their absolute magnitude sets the order of enumerated params in the output, aka "multis". Outputs tell which outputs of the scheme are produced. The numbers in this argument index the outputs slot of `filefmt.outputs`. To get the columns of the file where the data lives, we need to add the # of params, since those are listed in the first cols. Furthermore, `extract` can encode param values in cols of the output. The `colThresh` tells where to do this. It is a positive integer, we require `colThresh` <= the number of multis. For example, given pIndices = [1, -1, -3, 4, -2] and colThresh = 2. There are a total of 5 params+xtparams. We extract three of these, namely the ones at index 1, 2, and 4 in the concatentation of params and xtparams. In the output, we order the extracted dimensions with 1 in rows, 2 and 4 in cols, with 2 innermost. So if, for example, there are two outputs], param 2 has 3 elements and param 4 has 2 elements, then there will be 2*3*2 output columns. The order will be o02040, o12040, o02140, o12140, o02240, o12240, o02041, o12041, etc. The number of rows will equal the number of elts in params 1. Whether multiple files are involved depends on where the boundary of params and xtparams is. For example, if there are 3 params, then the -2 in pIndices concerns the 2nd xparam. We just assumed it has 2 elts so we'll be cycling between 2 files as we generate cols.