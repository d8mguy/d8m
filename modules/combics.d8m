// Combics provides a selection of useful combinatorial algorithms. This is intended to be an evolving module for quite some time.

export randperm, enumCombns, streamIndices

// Return a list(list(integer)) where each element is a length k list chosen from 0...n.
val enumCombns = \imp(n, k, totcount: integer) {
    if(k < 1 || k >= n) exit("bad args")
    var proto = (1..k).{index}
    var rslt: list(list(integer)) = [copy(proto)]
    val nkdiff = n - k
    var incrAt = k - 1
    // while condition is final exit criterion
    while(incrAt > 0 || proto[0] < n - k) {
        // loop through the options at current incrementing position
        while((incrAt < k - 1 && proto[incrAt] < proto[incrAt+1]+1) || proto[incrAt] < n - 1) {
            proto[incrAt] += 1
            rslt.pushb(copy(proto))
        }
        if(rslt.count > totcount) break
        // then reset incrAt and values in proto to set up next incrementing sequence
        while(incrAt > 0) {
            incrAt -= 1
            if(proto[incrAt] < nkdiff + incrAt) {
                var nval = proto[incrAt]+1
                each(h^incrAt...k) {
                    proto[h] = nval
                    nval += 1
                }
                rslt.pushb(copy(proto))
                incrAt = k - 1
                break
            }
        }
    }
    rslt
}

// A stream type that accepts a list of lengths and generates all the indices up to each of those lengths, in
// lexicographic order. Thus, given [1,2,3] it will generate [[0,0,0],[0,0,1],[0,0,2],[0,1,0],[0,1,1],[0,1,2]].
// Note that `[streamIndices: ]` is immediately done, it generates nothing. This is the right way to handle the
// situation where you need to enumerate all the elements of a variable number of lists. It's sensible to think of the k
// values that initialize a `streamIndices` as the extents of a k-dimensional cube. From this perspective, the
// stream generates all vertices of the cube in lexicographic order. A  `streamIndices` will panic if any initializing length
// is <= 0 or `next` is called when `done`.
//
// A `stream type` by convention has methods named `current`, `next`, and `done`; you call `next` until `done` becomes true.
val streamIndices = extend tuple(cur, limits: list(integer)) where {
    method lithook = \mod(lims: list(integer)) {
        if(any(lims.{this < 1})) exit("streamIndices requires positive limits")
        limits = lims.{this - 1}
        cur = zerolist(lims.count)
        unless(lims.count == 0) cur[lims.count - 1] = -1     // initialize so next returns all zeros first time
    }
    // utility method: number of elts produced by self from create to done
    method eltcount = \() { reduce(limits, \(a,b: integer) { a*(b+1)}, 1) }
    // utility method: number of "dimensions"
    method nDimensions = \() { cur.count }
    method restart = \mod() {
        cur = zerolist(limits.count)
        unless(limits.count == 0) cur[limits.count - 1] = -1
    }
    method done = \() { cur == limits }
    method current = \() { cur }
    method next = \mod() -> list(integer) {
        val nelts = limits.count
        each(i^1..nelts) {
            val ii = nelts - i; if(cur[ii] < limits[ii]) { cur[ii] += 1; each(j^(ii + 1)...nelts) cur[j] = 0; return cur }
        }
        exit("should not call terminated stream")
    }
}

// Randomly permute a list(T), using go package `math/rand` with implicit seed to generate the random numbers. It modifies and returns the list.
// Note that repeated calls to this function will not generate all its permutations without repetition.
// Note also that since it uses the PRNG state built into `math/rand`, using this may change
// the behavior of other program elements that also use that PRNG. If this is a problem, you can easily write
// a version with its own state; the code is just 8 lines long.
// The algorithm here is called the Fisher-Yates shuffle; as explained in the wikipedia, a version was
// published in 1938 by Fisher and Yates, and popularized in vol 2 of Knuth's TAOCP.

given(T::entity) {
    val randperm = \mod(lst: list(T), randSrc: \(integer)->integer) {
        var lim = count(lst)
        each(i^0...count(lst)-1) {
            val j = i+randSrc(lim)
            swap(lst, i, j)
            lim -= 1
        }
        lst
    }
}
