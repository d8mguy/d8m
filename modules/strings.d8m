/*
 * This module provides access to a subset of the functions in the corresponding go pkg, plus a few others.
 */
import go "strings" as gs

export split, join, toLower, toUpper, toCap, toCamel, genUpper, genLower, toOrdinal, startsWith, endsWith, trimSpace

val join = \(strgs: list(string), inter: string) -> string { gs.Join(strgs, inter) }
// Be careful: gs.Split doesn't return an empty string if no instances of sep in strg, it returns a singleton list of strg
// Thus, the correct test for no separators is split.count = 1, not 0.
val split = \(strg, sep: string) -> list(string) { gs.Split(strg, sep) }
val toLower = \(strg: string) -> string { gs.ToLower(strg) }
val toUpper = \(strg: string) -> string { gs.ToUpper(strg) }

val toCap = \imp(strg: string) -> string {
    var ret: list(byte) = strg
    if(strg != "") {
        val byt0 = strg[0]
        if('a' <= byt0 && byt0 <= 'z') ret[0] = 'A' + byt0 - 'a'
    }
    cvt(ret, string)
}

// Generate a camel case string from the input strgs; initial says whether to capitalize the first word.
val toCamel = \(strgs: list(string), initial: boolean) -> string {
    if(initial) join(strgs.{this.toCap}, "")
    else if(strgs.count == 1) strgs[0]
    else if(strgs.count == 0) ""
    else strgs[0] + join(tail(strgs).{this.toCap}, "")
}

// Generate letter sequence elements from indices for various underlying sets of letters
val genUpper = \imp(inx: integer) -> string {
    val letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var inx0 = inx
    var bytes: list(byte) = []
    while(inx0 > 0) { pushf(bytes, letters[inx0 % 26]); inx0 /= 26 }
    cvt(bytes, string)
}

val genLower = \imp(inx: integer) {
    val letters = "abdefghijklmnopqrstuvwxyz"
    var inx0 = inx
    var bytes: list(byte) = []
    while(inx0 > 0) { pushf(bytes, letters[inx0 % 26]); inx0 /= 26 }
    cvt(bytes, string)
}

// This isn't the only way to generate ordinals but it's reasonable.
val toOrdinal = \imp(xxx: integer) -> string {
    var x0 = xxx
    var prefix = ""
    if(x0 < 0) { x0 = -x0; prefix = "-" }
    var suffix = "th"
    if(x0 % 10 == 1 && x0 != 11) suffix = "st"
    else if(x0 % 10 == 2 && x0 != 12) suffix = "nd"
    prefix + "#{x0}" + suffix
}

// true if strg starts with ptn
val startsWith = \(strg, ptn: string) -> boolean {
    val cnt = strg.count
    if(ptn.count <= cnt) strg[0...cnt-ptn.count] == ptn
    else false
}

// true if strg ends with ptn
val endsWith = \(strg, ptn: string) -> boolean {
    val cnt = strg.count
    if(ptn.count <= cnt) strg[cnt-ptn.count...cnt] == ptn
    else false
}

// trim whitespace from both ends. This is done in a go pkg so we use that
val trimSpace = \imp(strg: string) -> string {
    gs.TrimSpace(strg)
}

