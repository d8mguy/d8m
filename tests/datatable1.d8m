// test the dataTable module with a single datascheme involving 4 params and 3 xparams. Use a synthetic file generator
// to get predictable values in the files, run 4 variants of the params to extract to test various cases.

import "dataTable" as csv
import "strings"
import "combics" melted

val params = [
    [csv.paramT: "P1", "", "", ["e00", "e01", "e02", "e03", "e04"], true, true],
    [csv.paramT: "P2", "", "", ["e10", "e11"], true, true],
    [csv.paramT: "P3", "", "", ["e20", "e21", "e22"], true, true],
    [csv.paramT: "P4", "", "", ["e30", "e31", "e32", "e33", "e34", "e35"], true, true]]
val xParams = [
    [csv.paramT: "X1", "", "", ["x00", "x01", "x02", "e03", "e04"], true, true],
    [csv.paramT: "X2", "", "", ["x10", "x11", "x12", "e13"], true, true],
    [csv.paramT: "X3", "", "", ["x20", "x21"], true, true]]
val outputs = ["o1", "o2", "o3", "o4"]
// For synth, the fnm is the wanted output for this "file"; synthDatatable1 is specialized for the above
// params; it returns a table of the 4 params and 4 outputs where the output values tell xparam values
val synthDatatable1 = \imp(fnm: string) -> csv.rowcolT {
    var rownum = 0
    var pStrm = [streamIndices: params.{ elts.count }]
    var rslt: csv.rowcolT = []
    while(!pStrm.done) {
        var prms = pStrm.next.{ params[index].elts[this] }
        val outs = outputs.{ strings.join([this, stringify(rownum), fnm], "/") }
        rslt.pushb(prms.append(outs))
        rownum += 1
    }
    rslt
}
val foo = \imp() {
    var state = [csv.datascheme: [csv.fileFormat: params, outputs], synthDatatable1]
    val fnames = map(i^0...5, j^0...4, k^0...2) strings.join([xParams[0].elts[i], xParams[1].elts[j], xParams[2].elts[k]], "")
    state.update(xParams, fnames, "")
    val rc0 = state.extract([2, -1, 2, 3, 0, -2, 0], [0], 2)
    println("state.extract([2, -1, 2, 3, 0, -2, 0], [0], 2):", rc0)
    val rc0b = state.extract([2, -1, 2, 3, 1, -2, 0], [0], 2)
    println("state.extract([2, -1, 2, 3, 1, -2, 0], [0], 2):", rc0b)
    val rc0c = state.extract([2, -1, 2, 3, 0, -2, 1], [0], 2)
    println("state.extract([2, -1, 2, 3, 0, -2, 1], [0], 2):", rc0c)
    val rc1 = state.extract([2, -1, 2, 3, 0, -2, 0], [0], 1)
    println("state.extract([2, -1, 2, 3, 0, -2, 0], [0], 1):", rc1)
    val rc2 = state.extract([2, -2, 2, 3, 0, -1, 0], [0], 2)
    println("state.extract([2, -2, 2, 3, 0, -1, 0], [0], 2):", rc2)
    val rc3 = state.extract([0, 1, -1, -2, 0, 2, 0], [1,2], 2)
    println("state.extract([0, 1, -1, -2, 0, 2, 0], [1,2], 2):", rc3)
}
foo()
