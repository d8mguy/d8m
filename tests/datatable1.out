state.extract([2, -1, 2, 3, 0, -2, 0], [0], 2): [[o1/87/x00x10x20], [o1/87/x00x11x20], [o1/87/x00x12x20], [o1/87/x00e13x20], [o1/105/x00x10x20], [o1/105/x00x11x20], [o1/105/x00x12x20], [o1/105/x00e13x20]]
state.extract([2, -1, 2, 3, 1, -2, 0], [0], 2): [[o1/87/x01x10x20], [o1/87/x01x11x20], [o1/87/x01x12x20], [o1/87/x01e13x20], [o1/105/x01x10x20], [o1/105/x01x11x20], [o1/105/x01x12x20], [o1/105/x01e13x20]]
state.extract([2, -1, 2, 3, 0, -2, 1], [0], 2): [[o1/87/x00x10x21], [o1/87/x00x11x21], [o1/87/x00x12x21], [o1/87/x00e13x21], [o1/105/x00x10x21], [o1/105/x00x11x21], [o1/105/x00x12x21], [o1/105/x00e13x21]]
state.extract([2, -1, 2, 3, 0, -2, 0], [0], 1): [[o1/87/x00x10x20, o1/87/x00x11x20, o1/87/x00x12x20, o1/87/x00e13x20], [o1/105/x00x10x20, o1/105/x00x11x20, o1/105/x00x12x20, o1/105/x00e13x20]]
state.extract([2, -2, 2, 3, 0, -1, 0], [0], 2): [[o1/87/x00x10x20], [o1/105/x00x10x20], [o1/87/x00x11x20], [o1/105/x00x11x20], [o1/87/x00x12x20], [o1/105/x00x12x20], [o1/87/x00e13x20], [o1/105/x00e13x20]]
state.extract([0, 1, -1, -2, 0, 2, 0], [1,2], 2): [[o2/18/x00x12x20, o3/18/x00x12x20], [o2/19/x00x12x20, o3/19/x00x12x20], [o2/20/x00x12x20, o3/20/x00x12x20], [o2/21/x00x12x20, o3/21/x00x12x20], [o2/22/x00x12x20, o3/22/x00x12x20], [o2/23/x00x12x20, o3/23/x00x12x20], [o2/24/x00x12x20, o3/24/x00x12x20], [o2/25/x00x12x20, o3/25/x00x12x20], [o2/26/x00x12x20, o3/26/x00x12x20], [o2/27/x00x12x20, o3/27/x00x12x20], [o2/28/x00x12x20, o3/28/x00x12x20], [o2/29/x00x12x20, o3/29/x00x12x20], [o2/30/x00x12x20, o3/30/x00x12x20], [o2/31/x00x12x20, o3/31/x00x12x20], [o2/32/x00x12x20, o3/32/x00x12x20], [o2/33/x00x12x20, o3/33/x00x12x20], [o2/34/x00x12x20, o3/34/x00x12x20], [o2/35/x00x12x20, o3/35/x00x12x20]]
